sounddevice.py

```diff
@@ -2349,6 +2399,157 @@ def __init__(self, channel_selectors):
+        # through kwargs.
+        del _
+        kind = kwargs['kind']
+        device = kwargs['device']
+        channels = kwargs['channels']
+        # This is to allow some flexibility for other Host-API-specific
+        # settings such as AsioSettings or WasapiSettings and not allow
+        # any accidental reliance on positional parameters.
+
+        if not isinstance(channels, int):
+            raise TypeError('channels must be an int, but found {!r} which is type {!r}'.format(channels, type(channels)))
+
+        if self._channel_selectors!=None and channels!=len(self._channel_selectors):
+            raise ValueError('channels is {}, but the length of channel_selectors is {}'.format(channels, len(self._channel_selectors)))
+
+        this_channelmap = None
+        if self._channel_selectors is not None:
+            # this array must be kept alive!
+            if kind=='input':
+                max_channels = query_devices(device, 'input')['max_input_channels']
+                for ndx in range(len(self._channel_selectors)):
+                    if self._channel_selectors[ndx]>=max_channels:
+                        raise ValueError("channel_selectors[{}]=={} exceeds max_channels=={}".format(ndx, self._channel_selectors[ndx], max_channels))
+                    if self._channel_selectors[ndx]<0:
+                        raise ValueError("channel_selectors[{}]=={} is < 0".format(ndx, self._channel_selectors[ndx]))
+                # No "inversion" is necessary for 'input' devices:
+                this_channelmap = self._channel_selectors
+            elif kind=='output':
+                # "invert" channel_selectors for Core Audio output
+                # devices, as described at
+                # https://lists.apple.com/archives/coreaudio-api/2004/Jan/msg00141.html
+                max_channels = query_devices(device, 'output')['max_output_channels']
+                this_channelmap = [-1 for _ in range(max_channels)]
+                for ndx in range(len(self._channel_selectors)):
+                    if self._channel_selectors[ndx]>=max_channels:
+                        raise ValueError("channel_selectors[{}]=={} exceeds max_channels=={}".format(ndx, self._channel_selectors[ndx], max_channels))
+                    if self._channel_selectors[ndx]<0:
+                        raise ValueError("channel_selectors[{}]=={} is < 0".format(ndx, self._channel_selectors[ndx]))
+                    # This is where we "invert" the channel mapping:
+                    this_channelmap[self._channel_selectors[ndx]]=ndx
+            else:
+                raise ValueError("kind must be either 'input' or 'output', but found {!r} instead".format(kind))
+            #print('Debug: kind={} max_channels={} channel_selectors={!r} this_channelmap={!r}'.format(kind, max_channels, self._channel_selectors, this_channelmap))
+
+        cache_key = kind,device,channels
+        if cache_key in self._cache_cffi_streaminfo:
```

Comment:
The "cache" is just in case someone reuses the same object for opening multiple, different devices, so that I don't accidentally permit a CFFI-allocated struct to be freed before the caller really is done with it.  Is there some guarantee that a library user cannot pass the same instance to different devices?

One use case for feeding the same MacCoreSettings struct to multiple devices would be if a caller wants to specify change_dev_params or allow_rate_conv.  But yeah, in the case of a caller also specifying channel_selectors, they're likely going to be using separate objects for separate devices to specify separate mappings.

In any case, I purposefully kept this cache only at the instance level so that memory will be freed as soon as the instance itself is freed, similar to AsioSettings and WasapiSettings.

I only bothered with this MacCoreSettings class to present the same API as the AsioSettings and WasapiSettings.  In my own opinion, I'd rather do away with all 3 of these API-specific settings classes in favor of having the caller pass in a simple dict as I mentioned above.  Then the CFFI (_streaminfo) structs could be allocated during _get_stream_parameters and freed as soon as RawStream/Stream/whatever is done with them.  I'm game to throw together a draft of this idea if folks are interested in reviewing it further.